File: Projects/leetcode/canPartition.java
package leetcode;

import java.util.Arrays;

class LeetCode {
    public static boolean canPartition(int[] nums) {
        final int sum = Arrays.stream(nums).sum();
        if (sum % 2 == 1) return false;
        final int target = sum / 2;
        final boolean[] cache = new boolean[target + 1];
        cache[0] = true;
        for (int i = 1; i <= nums.length; i++) {
            final int num = nums[i - 1];
            for (int j = target; j >= num; j--) {
                cache[j] = cache[j] || cache[j - num];
            }
        }
        return cache[target];
    }
}

File: Projects/leetcode/canPartition.kt
package leetcode

fun canPartition(nums: IntArray): Boolean {
    val sum = nums.sum()
    if (sum % 2 == 1) return false
    val target = sum / 2
    val cache = BooleanArray(target + 1)
    cache[0] = true
    for (i in 1..nums.size) {
        val num = nums[i - 1]
        for (j in target downTo num) {
            cache[j] = cache[j] || cache[j - num]
        }
    }
    return cache[target]
}
